packages/server/lib/browsers/firefox.ts

```diff
@@ -458,7 +458,10 @@ export async function open (browser: Browser, url: string, options: BrowserLaunc
   const [
     foxdriverPort,
     marionettePort,
-  ] = await Promise.all([getPort(), getPort()])
+  ] = await Promise.all([
+    Number(process.env.CYPRESS_FOXDRIVER_PORT) || getPort(),
```

Comment:
>> These port values can't be managed today without the ENVS because we aren't honoring and pulling these values back out from the user defined options. If we can pull these out from the [launchOptions](](https://github.com/cypress-io/cypress/compare/emily/ff-dynamic-marionette-port#diff-6ec303ebac972615597a2bbcc7cd2986c44cea3ff5b0eccb0b8724a8527d1972R566)https://github.com/cypress-io/cypress/compare/emily/ff-dynamic-marionette-port#diff-6ec303ebac972615597a2bbcc7cd2986c44cea3ff5b0eccb0b8724a8527d1972R468) we can pass them into [firefox.setup()`, which would override the dynamic port allocation we are doing.

>> This would allow for:
>> ```js
>>     setupNodeEvents (_on, config) {
>>       _on('before:browser:launch', (browser, opts) => {
>>         if (process.env.CYPRESS_FOXDRIVER_PORT) {
>>           opts.preferences['devtools.debugger.remote-port'] = process.env.CYPRESS_FOXDRIVER_PORT
>>         }
>> 
>>         if (process.env.CYPRESS_MARIONETTE_PORT) {
>>           opts.preferences['marionette.port'] = process.env.CYPRESS_MARIONETTE_PORT
>>         }
>> 
>>         return opts
>>       })
>>   },
>> ```

> I am not sure I fully understand this last part (I think the link may have been miscopied?) but if I understand the intention of the example this does look promising. Being able to override devtools.debugger.remote-port would suffice for me if the other two ports could be dynamically allocated by Firefox itself.

Sorry for this confusing message -- rereading it took me a minute to figure out what the heck I was saying....

Reply:
see: [`emily/ff-dynamic-marionette-port`#diff-6ec303ebac](https://github.com/cypress-io/cypress/compare/emily/ff-dynamic-marionette-port#diff-6ec303ebac972615597a2bbcc7cd2986c44cea3ff5b0eccb0b8724a8527d1972R586)